### ISSUE TITLE
Feature Request: Metrics Directory Option

### ISSUE THREAD

**Issue by therealbill** [fstab/grok_exporter#76]
First, I'll clarify that this is not about just the ability reading multiple log files. I'll describe the use case to explain it.

Say you have a few hundred servers of various purposes (roles). You've got a common set of matches:metrics you want across all of them (such as matching things in systemd for example). For that purpose a single config file with all match/metrics defined is fine.

But say on some hosts you have log files and metrics you need to implement. Rewriting the config file on host creation is not a scalable option. Instead, if the exporter took a "-metrics_dir" option and it then read .yml files from there and integrated them internally into the metrics structure. Then from the operational perspective you can conditionally drop the files in there in w/e software packaging/deployment method you used and know they'd be used but only on those servers.

A few examples of this paradigm in-use would be Apache and Consul.

For grok, I'd look for something like:

in config:
```
grok:
  patterns_dir: /etc/grok_exporter/patterns
  metrics_dir: /etc/grok_exporter/metrics.d
```

Which could have the following files:
/etc/grok_exporter/metrics.d/systemd.yml
/etc/grok_exporter.d/metrics.d/consul.yml
/etc/grok_exporter.d/metrics.d/customserver.yml

These files would include the matches listing/definition normally found in the config, and these would be merged by grok. Basically each of those would look something like this:

```
- type: counter
  name: consul_error_logs
  help: number of log lines containing an error in consul.log
  match: '%{DATESTAMP} \[ERR\] *''
  path: /var/log/consul.log
  labels:
      logfile: '{{base .logfile}}'
```

Think of it like the metrics version of the patterns_dir. With this, your named hosts can have named-specific patterns and metrics, your Postfix hosts can have Postfix-specific patterns and metrics, etc. - all while having a common set they all need and without needing to write a large config file that has to have everything in it or having to use the option to ignore lack of files across the board because not all hosts will have all files.

**Comment by fstab** [comment]
I like the idea, but I'm not sure how a good config file structure would look like. The new `metrics_dir` key belongs to the `metrics` section, just like the `patterns_dir` is in the `grok` section. However, the `metrics` section is currently a list:

```yaml
# current metrics section
metrics:
    - type: counter
      name: errors_total
      help: Total number of error messages.
      match: 'ERROR'
    - type: counter
      name: errors_total
      help: Total number of error messages.
      match: 'ERROR'
```

We cannot add a key to a list. The following is invalid YAML:

```yaml
# invalid yaml
metrics:
    metrics_dir: /path/to/metrics.d
    - type: counter
      name: errors_total
      help: Total number of error messages.
      match: 'ERROR'
    - type: counter
      name: warnings_total
      help: Total number of warn messages.
      match: 'WARN'
```

In the `grok` section, we have this:

```yaml
# current grok section
grok:
    patterns_dir: ./logstash-patterns-core/patterns
    additional_patterns:
    - 'EXIM_MESSAGE [a-zA-Z ]*'
```

If we structure the `metrics` section like the `grok` section, we need to add one more level:

```yaml
# metrics section with similar structure as grok section
metrics:
    metrics_dir: /path/to/metrics.d
    additional_metrics:
        - type: counter
          name: errors_total
          help: Total number of error messages.
          match: 'ERROR'
        - type: counter
          name: warnings_total
          help: Total number of warn messages.
          match: 'WARN'
```

Assuming that most people won't use the `metrics_dir`, the most common config would then end up with a nested `metrics.additional_metrics`, which looks like an unnecessary indirection:

```yaml
# looks too nested
metrics:
    additional_metrics:
        - type: counter
          name: errors_total
          help: Total number of error messages.
          match: 'ERROR'
        - type: counter
          name: warnings_total
          help: Total number of warn messages.
          match: 'WARN'
```
Maybe we can move the config directories into the `global` section like this:

```yaml
# maybe a way to go forward
global:
    config_version: 2
    config_dirs:
      - grok_patterns: /etc/patterns.d
      - metrics: /etc/metrics.d
```

Then the `metrics` section can stay as it is, but we need to remove `patterns_dir` from the `grok` section. Maybe we can make the `grok` section a list, so that it's more aligned with the `metrics` section:

```yaml
# grok section as a list without patterns_dir
grok:
    - name: EXIM_MESSAGE
      pattern: '[a-zA-Z ]*'
```
I don't know what's the best way forward, need to sleep over it. If you have suggestions you are welcome to comment.

**Comment by therealbill** [comment]
Yes I agree that a nested addition_metrics indirection is unnecessary. I'm of a few minds on it, so I'll expand on them.

I'd go with using the same basic mechanism as patterns_dir. What I'd do is, in the code (my apologies, I haven't looked yet so I am unable to be specific at the moment) at the point where you read in the additional_metrics, I would do to the following:

0. Initialize the data structure that will contain the list of defined metrics
1. Consult the metrics_dir
2. Iterate over each file in there, perhaps in alphanumeric order, and read them as a list of individual file.
3. Parse each file as it's own list of defined metrics, and add each one to the metrics collection data structure
4. Read in "addition_matches" and add them in
5. Move along

The advantage to this is that it is programmatically quite simple, and you're pretty much re-using the metrics definition parsing code. I imagine you are doing that same sequence on patterns.

Now as to the larger config file discussion you started. ;)

Part of me likes the clarity of a metrics section. On the other hand, I also quite like consistency of form and behavior in a config, so I feel like the patterns and the metrics are two sides of the same coin and thus the overall config would benefit from being the same setup. It is also a lot of work that I don't think gets you much compared to the simpler route of mirroring the pattern of the patterns config.

If we look at my sequence above, the metrics section in the main config would be the counterpart to "additional_matches" where you can define new patterns in there that could easily be in a dedicated patterns file in the patterns directory. Another benefit is that your config structure remains backward compatible if/when you/we add this.

In the longer run, I'd probably look at deprecating the additional patterns and metrics in the main config file and go fully to the config directory route. It simplifies the config headspace for the user and provides the ability for operators and software authors to package and distribute patterns and metrics. 

Now on a slight tangent, and probably a separate request but related so I'll mention it, it would be stellar if adding a path/paths statement to a metric also added it (internally) to the list of files to monitor. As it is now you have to add it to two places in the config (two sections - the main source definition and each metric that is restricted to a specific file). Being able to say "here is a file that has metric definitions, they are limited to this log source" and having that be the way to define sources is much more convenient. On further thought last night it would also be needed to really meet the scenario I'm looking at.

That brings me to a related side question I haven't got around to trying yet: can I specify both stdin and files to follow?

I'll dig into the code a bit over the next several days to get a better understanding of how you've structured and and possibly can help. I like the basic idea of grok and think that with a few, limited, additions it could be an excellent tool. Then again, log analysis and parsing has been a passion of mine for over a decade. ;) I can see myself writing several pattern/metrics sets for things like systemD, Postfix, and others, but to be operationally solid across machines, the ability to have isolated configs is a game changer - in a good way.

**Comment by therealbill** [comment]
Ok, done a bit of digging, and this is roughly what I am thinking:

1. Add the metrics_dir to GrokConfig 
2. Add function to MetricsConfig to read and parse metrics in from directory something like "ReadMetricsDir"
3. During config init [just after addDefaults](https://github.com/fstab/grok_exporter/blob/master/config/v2/configV2.go#L123), call MetricsConfig's ReadMetricsDir with the directory specified in the config - if found. 

I may try a rough proof of concept out this weekend for reference, but that seems to me the most simple and direct. It also lets you keep actually [creating the metrics](https://github.com/fstab/grok_exporter/blob/5dda31958dddbb05f71f2580b1b89d376687ecde/grok_exporter.go#L214) the same as it is now.

Of course, I am probably missing something, but it looks relatively straightforward so far. 

Another way, but more invasive, would be to make [createMetrics](https://github.com/fstab/grok_exporter/blob/5dda31958dddbb05f71f2580b1b89d376687ecde/grok_exporter.go#L214) look and act like [initPatterns](https://github.com/fstab/grok_exporter/blob/5dda31958dddbb05f71f2580b1b89d376687ecde/grok_exporter.go#L197) where you already do it for patterns. Indeed, I suspect that would be the route I'd prefer to take if it were me.

**Comment by fstab** [comment]
Thanks for your POC. One of the goals should be that the `grok` section and the `metrics` section should be analogous. If you can specify both a `patterns_dir` and `additional_patterns` in the `grok` section, then you should be able to specify a `metrics_dir` and `additional_metrics` in the `metrics` section. Therefore I think we should not add `metrics_dir` to the `grok` section.

Another idea would be to create a whole new section `additional_config_files` that can be used for both, `grok_patterns` and `metrics`. An example could look like this:

```yaml
global:
    config_version: 3
input:
    type: file
    path: /var/log/syslog/*
grok_patterns:
    - 'FPMDATE %{MONTHDAY}/%{MONTH}/%{YEAR}:%{TIME} %{ISO8601_TIMEZONE}'
metrics:
    - type: counter
      name: no_of_segfaults
      help: Example counter metric with labels.
      match: "%{SYSLOGPROG}: segfault at %{BASE16NUM} ip %{BASE16NUM} sp %{BASE16NUM} error %{NUMBER} in %{GREEDYDATA}"
additional_config_files:
    - type: metrics
      dir: /etc/grok_exporter/metrics.d/
      defaults:
          paths: /var/log/syslog/*
          labels:
              source: syslog
              logfile: '{{base .logfile}}'
    - type: grok_patterns
      dir: /etc/grok_exporter/patterns/
server:
    host: 0.0.0.0
    port: 9144
```

The `defaults` could be used to specify default values applying to all metrics loaded from that config dir. That way we provide an easy way to add a `paths` statement to all metrics loaded from a config dir. As an alternative to `dir` we could also allow `files` with a Glob pattern, like `files: /etc/grok_exporter/metrics.d/*-syslog.yaml`.

Regarding your other question:
> Can I specify both stdin and files to follow?

No, there can only be one input, which is either `file` or `stdin` or `webhook`.

**Comment by fstab** [comment]
I created an `import-metrics-config` branch. An example config looks like this:
```yaml
global:
  config_version: 3
input:
  type: file
  path: ./example/exim-rejected-RCPT-examples.log
  readall: true
imports:
- type: grok_patterns
  dir: ./logstash-patterns-core/patterns
- type: metrics
  file: ./example/metrics.yaml
grok_patterns:
- EXIM_MESSAGE [a-zA-Z ]*
server:
  protocol: http
  host: localhost
  port: 9144
```

The imported `metrics.yaml` looks like this:
```yaml
- type: counter
  name: exim_rejected_rcpt_total
  help: Total number of rejected recipients, partitioned by error message.
  match: 'WARNING'
  labels:
    error_message: '{{.message}}'
    logfile: '{{base .logfile}}'
```
The currently supported fields in `imports:` are here:
https://github.com/fstab/grok_exporter/blob/2027b098438c1ca63688d10d4c5b85cb967a802c/config/v3/configV3.go#L120-L133

You can import a single file or a glob pattern as in the example, but there is also a `dir:` field to import all files in a directory.

**Comment by fstab** [comment]
I'm done with the implementation. Updated documentation is on the feature branch [github.com/fstab/grok_exporter/blob/import-metrics-config/CONFIG.md](https://github.com/fstab/grok_exporter/blob/import-metrics-config/CONFIG.md). I will soon merge it to master and create a new release.

**Comment by fstab** [comment]
Released it with [v1.0.0.RC3](https://github.com/fstab/grok_exporter/releases/tag/v1.0.0.RC3). If you have some metrics configs for things like systemD, Postfix, and others I'm happy to merge them. Thanks for the idea.
